### PR TITLE
Mission inventory: filter in the store, not after the fact

### PR DESCRIPTION
**Pull request by Th0rgal**
First PR of the project ↔ conversation ↔ missions series. It makes the inventory truthful, which everything else depends on.

## The bug

`GET /api/control/missions?project=…` ran its predicate **in memory** over paged reads, giving up after 5 000 rows. On a fleet this size that horizon is reached well inside a day, so a conversation whose missions had scrolled past it showed **zero workers — silently, with a 200 OK**. Both clients hit it: `HermesConversation.tsx` and iOS `HermesConversationView` each fetched the newest 50 missions and filtered client-side.

## Changes

- **`MissionFilter`** is now the single source of predicate truth (`status`, `project`, `project_prefix`, `track`, `tag`, `origin_session_id`). The in-memory scan and the SQL pushdown call the same `matches()`, so they cannot drift apart on what a filter means.
- **`MissionStore::list_missions_filtered`** — default impl keeps today's bounded scan (file/memory stores); **sqlite overrides it with a real `WHERE`** and is exact at any depth.
- **`project` stays exact.** Cross-phase queries use `project_prefix`, a hyphen-anchored family match: `verity` covers `verity-core` / `verity-phase1d`, but never `verityx` or `x-verity`. Redefining `project` itself would have silently widened results for every existing caller (dashboard, iOS, Paloma).
- **`track` and `origin_session_id` filters** — the latter is what the Hermes conversation view actually needs.
- **`tag`** is a SQL `LIKE` prefilter plus a Rust re-check, deliberately **not `json_each`**: `tags` is JSON TEXT and some rows hold malformed blobs, which would fail the entire query. There's a test with exactly such a row next to the match.

## Two traps worth reviewing

1. **Index placement.** The two new indexes are created at the **end of `run_migrations`**, never in `SCHEMA`. `execute_batch(SCHEMA)` runs first, and `CREATE TABLE IF NOT EXISTS missions` is a no-op on an existing DB — so indexing a column the ALTERs have not added yet would fail store init and **silently drop the service to an in-memory store**, the same trap the ALTER block above already documents. Non-fatal handling to match.
2. **Shared projection.** The row→`Mission` mapping and its column list are now one `row_to_mission` + one `MISSION_LIST_COLUMNS`. The getters are positional, so two copies of a 60-line projection were a drift waiting to happen.

## Testing

`cargo test --lib` — 1640 passed, including 5 new sqlite filter tests:
- a match buried under a large fleet is still returned (**the regression**);
- family match hits `verity`/`verity-phase1d` and not `verityx`/`x-verity`;
- `project` stays exact;
- `offset` skips **matches**, not raw rows;
- tag filter survives a malformed `tags` blob on a neighbouring row.

Dashboard: 3 new vitest cases on the built query string; `cargo fmt --all` clean.

## Follow-ups in this series
Short-id resolution + `origin` in the MCP summary (PR B), the explicit project→conversation binding (PR C), then the Hermes side.

---

> [!NOTE]
> **Medium Risk**
> Touches core mission inventory API and SQLite listing paths used by dashboard, iOS, and MCP; behavior changes are intentional but `project` semantics must stay exact while new filters roll out.
> 
> **Overview**
> Fixes **Hermes conversations showing no workers** when spawned missions had aged past the default “newest 50” page. Listing used to filter in memory over paged reads (capped around 5k rows), so matches could disappear with a normal **200 OK**.
> 
> **Backend:** Introduces shared **`MissionFilter`** and **`list_missions_filtered`**, with SQLite pushing predicates into **`WHERE`** (exact `project`, hyphen-anchored **`project_prefix`**, **`track`**, **`tag`**, **`origin_session_id`**). Workspace filtering stays in-memory but runs on store-filtered pages. Adds migration-time indexes on **`origin_session_id`** and **`(project, track)`**, plus consolidated **`row_to_mission`** / **`MISSION_LIST_COLUMNS`**.
> 
> **Clients:** Web **`listMissions`** options and iOS **`APIService.listMissions`** pass query params; Hermes conversation views request **`origin_session_id`** server-side instead of client-side filtering. Vitest covers the built query string; new SQLite filter tests cover the buried-match regression and edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit c6884d6198b9e991bc1190338171f22b160f0dc2. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: dashboard/src/app/control/components/HermesConversation.tsx [hunk 1]
```diff
@@ -304,11 +304,15 @@ export function HermesConversation({ sessionId }: { sessionId: string }) {
...
       }
       try {
-        const missions = await listMissions();
+        // Server-side filter: fetching the newest 50 and filtering here used
+        // to show zero workers for any conversation whose missions had
+        // scrolled out of that page.
+        const missions = await listMissions({
+          originSessionId: sessionId,
+          limit: 200,
+        });
```
[the remark this answers is not in the document]
Comment (Th0rgal) [reply dashboard/src/app/control/components/HermesConversation.tsx:313]: Fixed in e06b78f8 — you're right that leaving iOS on the client-side filter left the same user-visible bug in place. APIService.listMissions now takes origin_session_id/project/project_prefix/track/limit and builds the query with URLComponents; HermesConversationView passes the session id and drops its local filter.

File: src/api/mission_store/sqlite.rs [hunk 2]
```diff
@@ -2841,6 +2875,114 @@ impl MissionStore for SqliteMissionStore {
...
+            if let Some(v) = filter.track.as_deref() {
+                bind("track = ?n", v, &mut clauses, &mut binds);
+            }
+            if let Some(v) = filter.origin_session_id.as_deref() {
+                bind("origin_session_id = ?n", v, &mut clauses, &mut binds);
+            }
+            if let Some(v) = filter.tag.as_deref() {
+                bind(
+                    r#"tags LIKE '%"' || ?n || '"%'"#,
+                    v,
```
[the remark this answers is not in the document]
Comment (Th0rgal) [reply src/api/mission_store/sqlite.rs]: Fixed in e06b78f8. The prefilter now matches the tag's JSON serialization (serde_json::to_string, quotes included) instead of raw text, so an escaped quote/backslash/newline is no longer excluded in SQL before MissionFilter::matches can see it. Over-matching stays harmless — LIKE wildcards inside a tag only widen the candidate set and the Rust re-check is still the authority. Added tag_filter_matches_json_escaped_tags covering all three shapes.

File: src/api/mission_store/sqlite.rs [hunk 3]
```diff
@@ -2841,6 +2875,121 @@ impl MissionStore for SqliteMissionStore {
...
+            if let Some(v) = filter.tag.as_deref() {
+                // Match the tag's JSON *serialization*, quotes included: a tag
+                // holding a quote/backslash/newline is escaped inside the
+                // column, so a raw-text LIKE would exclude it in SQL and the
+                // Rust re-check below would never get to see it. Over-matching
+                // is fine here (LIKE wildcards in the tag only widen the
+                // candidate set); under-matching would be a silent miss.
+                let encoded = serde_json::to_string(v).unwrap_or_else(|_| format!("\"{v}\""));
+                bind(
+                    "tags LIKE '%' || ?n || '%'",
```
[the remark this answers is not in the document]
Comment (Th0rgal) [reply src/api/mission_store/sqlite.rs]: Fixed. You're right and my 'over-matching is harmless' note was wrong — the cost isn't wasted rows, it's paging them under the connection lock. Now escaping \, % and _ in the encoded pattern (escape char first so backslashes don't double) with ESCAPE '\'. Added tag_filter_treats_like_wildcards_as_literals: a bare '%' matches nothing, and a tag genuinely containing '%' is still found.

File: src/api/mission_store/sqlite.rs [hunk 4]
```diff
@@ -2841,6 +2875,129 @@ impl MissionStore for SqliteMissionStore {
...
+                };
+            if let Some(v) = filter.status.as_deref() {
+                bind("status = ?n", v, &mut clauses, &mut binds);
+            }
+            if let Some(v) = filter.project.as_deref() {
+                bind("project = ?n", v, &mut clauses, &mut binds);
+            }
+            if let Some(v) = filter.project_prefix.as_deref() {
+                bind(
+                    "(project = ?n OR project LIKE ?n || '-%')",
```
[the remark this answers is not in the document]
Comment (Th0rgal) [reply src/api/mission_store/sqlite.rs]: Fixed. Replaced the LIKE with range bounds: `project = family OR (project >= 'family-' AND project < 'family.')` — '-' is 0x2D and '.' is 0x2E, so the half-open interval is exactly the family-* set under BINARY collation and both branches can use idx_missions_project_track. Added project_family_lookup_uses_an_index, which asserts EXPLAIN QUERY PLAN contains no 'SCAN missions'; this regression class is invisible in behaviour so the plan itself needed pinning.

File: src/api/control/mod.rs [hunk 5]
```diff
@@ -4895,18 +4907,27 @@ pub async fn list_missions(
...
+    // Workspace is the one filter the store cannot answer: matching by name
+    // needs `populate_workspace_names`, and the persisted `workspace_name`
+    // column is not authoritative. So it keeps the in-memory scan.
+    let needs_workspace_scan = query.workspace.is_some();
     let limit = query.limit.unwrap_or(50).clamp(1, 200);
     let offset = query.offset.unwrap_or(0);
 
-    let mut missions = if !has_filters {
-        // Fast path: no filters, list the page directly.
+    let mut missions = if !needs_workspace_scan {
```
[the remark this answers is not in the document]
Comment (Th0rgal) [reply src/api/control/mod.rs:4928]: Fixed. The workspace branch now pages list_missions_filtered instead of list_missions, so the store predicate still runs in SQL and only the workspace-name match happens in memory. The 5 000 cap now bounds how many store-filtered candidates we resolve names for, not how deep we read — a selective filter means the store already skipped everything irrelevant.

**Comment by Th0rgal** [comment]
Fixed in #733. The finding was correct: `EXPLAIN QUERY PLAN` on the emitted track-only query reported `SCAN missions USING INDEX idx_missions_updated_at`. Added `idx_missions_track_updated ON missions(track, updated_at DESC)`, which turns it into `SEARCH missions USING INDEX idx_missions_track_updated (track=?)`, and verified the family / exact-project / origin_session plans are unchanged.
